release-22.2: kvserver: register lease renewals on splits/merges

erikgrinaker opened the pull request:
Backport:
  * 1/1 commits from "kvserver: register lease renewals on splits/merges" (#100392)
  * 2/2 commits from "kvserver: deflake `TestLeaseRenewer`" (#101012)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: fixes bugs that could prevent eager renewal of meta/liveness leases.

erikgrinaker committed:
kvserver: register lease renewals on splits/merges

The store lease renewer, which eagerly extends expiration-based leases,
did not correctly (un)register ranges on splits/merges/relocations. This
could cause the lease renewer to not eagerly extend leases for some
ranges, but was mitigated by the lease renewer only being used for the
meta and liveness ranges which rarely/never split.

This patch makes sure the lease renewer is notified on these events. It
also incidentally fixes an unlikely race where a range split immediately
following an expiration-based lease transfer could leave the RHS using
an expiration-based lease until it was transferred elsewhere.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100631)
<!-- Reviewable:end -->

erikgrinaker committed:
kvserver: deflake `TestLeaseRenewer`

Epic: none
Release note: None

erikgrinaker committed:
kvserver: fix lease renewal registration of expired leases

The store lease renewer begins tracking an system range's
expiration-based lease for renewal when it's first applied. However, it
only did this if the lease was still valid when first applied. It was
possible for the node to apply a lease after it's lapsed, but then to
retain and extend the lease, and in this case it would not be registered
with the lease renewer.

Epic: none
Release note: None

andrewbaptist reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
